### ISSUE TITLE
Restarts too many OSDs too often

### ISSUE THREAD

**Issue by mcv21** [ceph/ceph-ansible#1375]
Hi,
[using stable-2.1 branch]

We have a medium-sized ceph cluster, with 540 OSDs. We have mons, osds, rgws. Currently, a run of ansible-playbook will restart every osd multiple times. Given a restart takes ~80 minutes, this is deeply tedious. At least ceph-ansible should only restart osds on the hosts its currently working on (maybe using `play_hosts` or `ansible_play_batch` or similar?)

The current behaviour might also be harmful. First play is mons; if that results in changes to ceph.conf, then all the osds are restarted, including on those nodes which are not mons, so won't have had any ceph.conf changes applied to them yet.

**Comment by leseb** [comment]
What do you mean by restarting an OSD multiple times? How many times? If there is a config change OSD are restarted one by one and machine by machine. The restart is being serialized so I'm a bit surprised by your issue.

Can you share some logs?

It is difficult for us now to determine which option changed so we can restart the impacted daemons accordingly. This requires work on both ceph-ansible and ceph itselft. There is an upstream PR that should help us: https://github.com/ceph/ceph/pull/10736

Let me have another look into this with @ErwanAliasr1 for ceph-ansible part.

**Comment by mcv21** [comment]
Sorry for being unclear. Let me try again :)
Suppose I change something that will adjust ceph.conf; then what happens when I run ansible-playbook is:

- ansible runs the mons role, which changes ceph.conf on the mons. That notifies the restart ceph osds handler, which restarts osds across the entire cluster (not just the mons)
- ansible runs the osds role, which changes ceph.conf on servers with just osds. That notifies the restart ceph osds handler, which restarts osds across the entire cluster
- ansible runs the rgws role, which changes ceph.conf on the rgs. That notifies the restart ceph osds handlers, which restarts osds across the entire cluster (not just the rgws)

[a log of the last ansible-playbook run is 14k lines...]

**Comment by leseb** [comment]
Hum weird I thought we were safe because we use conditions like`mon_group_name in group_names` on tasks which should only restart services during each host play.

**Comment by mcv21** [comment]
I think it's because the handler has:

```  with_items: "{{ groups[osd_group_name] }}"
  delegate_to: "{{ item }}"
```

...so it always runs against every host in the `osds` group?

**Comment by guits** [comment]
could you provide a link to the log of your ansible-playbook run anyway?

**Comment by leseb** [comment]
@mcv21 Ansible version?

**Comment by mcv21** [comment]
`ansible 2.2.1.0`

**Comment by mcv21** [comment]
http://www.chiark.greenend.org.uk/~matthewv/pastes/ansibleout

Is a playbook log I kept (the ceph-mon run didn't change ceph.conf that time, so the total-restart only happens twice). I removed a couple of local plays (adding ssh keys etc).

**Comment by guits** [comment]
@mcv21 can you provide a link to your inventory host file?

**Comment by mcv21** [comment]
It's a dynamic inventory - 
```python
#!/usr/bin/env python

import json
import itertools
import string
import argparse

#a lazy way of producing 1-1 to 3-3
hostnames=["sto-%d-%d" %(x,y) for x,y in itertools.product(range(1,4),repeat=2)]

#the -1 host in each cab is a monitor
monhosts = filter(lambda x: x[-1]=="1", hostnames)

#the -2 host in each cab is a radosgw
rgwhosts = filter(lambda x: x[-1]=="2", hostnames)


hostvars = {}

#We need IP addresses in the 2 ceph subnets:
# 172.27.6.0/23 for the ceph "public" network
# 172.27.8.0/23 for the ceph internal network
# Arbitrarily, start at 11 (for sto-1-1) and work upwards.

n = 11
for h in hostnames:
    #create dictionary for first entry
    hostvars[h] = {"cephpubip" : "172.27.6.%d" % n}
    hostvars[h]["cephprivip"] = "172.27.8.%d" %n
    #sto-X-Y is in cabinet/rack X
    hostvars[h]["ceph_crush_rack"] = (h.split('-')[1])
    n+=1

## Declare groups and their top-level structure

#Variables for the OSDs
osds = {}
osds["hosts"] = hostnames
osds["vars"] = {}

#Variables for the Mons
mons = {}
mons["hosts"] = monhosts
mons["vars"] = {}

#Variables for the rados gateways
rgws = {}
rgws["hosts"] = rgwhosts
rgws["vars"] = {}

#Full output
inventory={"_meta": {"hostvars":hostvars},
           "osds": osds,
           "mons": mons,
           "rgws": rgws
       }

## OSDs Configuration

#We might only have to set rack= as the others are what ceph does by default
#http://docs.ceph.com/docs/jewel/rados/operations/crush-map/ 
#is a bit ambiguous, though, so set them here anyway
osds["vars"]["osd_crush_location"]="root=default rack={{ ceph_crush_rack }} host={{ ansible_hostname }}"
osds["vars"]["crush_location"]=True

#We might want to do something more clever re fsid; currently it is 
#generated on the ansible-host in fetch/ceph_cluster_uuid.conf
osds["vars"]["fsid"]="{{ cluster_uuid.stdout }}"
mons["vars"]["fsid"]="{{ cluster_uuid.stdout }}"

#devices - we want /dev/sda -> /dev/sdbh which should be 60 devices
#build this by taking lists a-z aa-az ba-bz and taking the first 60.
#and then appending to /dev/sd
atoz=list(string.ascii_lowercase)
aatoaz=map(lambda x: "a"+x, atoz)
batobz=map(lambda x: "b"+x, atoz)
devsuffs=(atoz+aatoaz+batobz)[:60]
osds["vars"]["devices"]=map(lambda x: "/dev/sd"+x, devsuffs)

#half the journals go on nvme0n1, the other half on nvme1n1
jourdevs=[]
for x in range(30):
    jourdevs.append("/dev/nvme0n1")
    jourdevs.append("/dev/nvme1n1")
osds["vars"]["raw_journal_devices"]=jourdevs
osds["vars"]["raw_multi_journal"] = True

##Mons configuration

mons["vars"]["monitor_secret"]="{{ monitor_keyring.stdout }}"

#Pool setup
#This is per https://access.redhat.com/documentation/en/red-hat-openstack-platform/9/single/red-hat-ceph-storage-for-the-overcloud/#config-ceph-cluster
# ceph-ansible has openstack config options, but they're really just
# general-purpose pool creation. That suits us fine.
# We want 3 pools for ceph:
# delta-volumes, delta-images, delta-vms
# And more pools for the radosgw:
#
#    default.rgw
#    default.rgw.control
#    default.rgw.gc
#    default.rgw.log
#    default.rgw.intent-log
#    default.rgw.usage
#    default.rgw.users
#    default.rgw.users.email
#    default.rgw.users.swift
#    default.rgw.users.uid
#    .rgw.root (could tweak name via rgw zonegroup root pool)
# Using pgcalc, and a rough spread of what we wanted usage-size, we want
# 128 for the small pools
# 256 for the 1% ones - rbd, default.rgw.buckets.extra, default.rgw.buckets.index
# 1024 for images
# 4096 for default.rgw.buckets and volumes
# 8192 for vms
# (the CRUSH map is slightly more efficient with power-of-2 sizes)

pools = []
for p in ["default.rgw", "default.rgw.control", "default.rgw.gc", \
          "default.rgw.log", "default.rgw.intent-log", "default.rgw.usage", \
          "default.rgw.users", "default.rgw.users.email", \
          "default.rgw.users.swift", "default.rgw.users.uid", ".rgw.root"]:
    pools.append({"name": p, "pg_num": 128})

for p in ["rbd", "default.rgw.buckets.extra", "default.rgw.buckets.index"]:
    pools.append({"name": p, "pg_num": 256})

pools.append({"name": "delta-images", "pg_num": 1024})

for p in ["default.rgw.buckets.data", "delta-volumes"]:
    pools.append({"name": p, "pg_num": 4096})

pools.append({"name": "delta-vms", "pg_num": 8192})

mons["vars"]["openstack_pools"] = pools

# We also want a client.openstack-delta user with permissions:
#  cap_mon: allow r
#   cap_osd: allow class-read object_prefix rbd_children, allow rwx pool=volumes, allow rwx pool=vms, allow rwx pool=images 

keys = []
k = {}
k["name"]="client.openstack-delta"
k["value"]="mon 'allow r' osd 'allow class-read object_prefix rbd_children, allow rwx pool=delta-volumes, allow rwx pool=delta-vms, allow rwx pool=delta-images'"
keys.append(k)

kdemo = {}
kdemo["name"]="client.openstack-demo"
kdemo["value"]="mon 'allow r' osd 'allow class-read object_prefix rbd_children, allow rwx pool=demo-volumes, allow rwx pool=demo-vms, allow rwx pool=demo-images'"
keys.append(kdemo)

kdelta2 = {}
kdelta2["name"]="client.openstack-delta2"
kdelta2["value"]="mon 'allow r' osd 'allow class-read object_prefix rbd_children, allow rwx pool=delta2-volumes, allow rwx pool=delta2-vms, allow rwx pool=delta2-images'"
keys.append(kdelta2)

mons["vars"]["openstack_keys"] = keys

mons["vars"]["openstack_config"] = True

## UI (such as it is)

def parse_args():
    parser = argparse.ArgumentParser(description='Ceph dynamic inventory')    
    group = parser.add_mutually_exclusive_group(required=True)
    group.add_argument('--list', action='store_true',
                       help='list all hosts')
    group.add_argument('--host', help='list a specific host')
    return parser.parse_args()

def main():
    args = parse_args()
    if args.list:
        print json.dumps(inventory,indent=2,sort_keys=True)
    elif args.host:
        print json.dumps(hostvars[args.host])

if __name__ == "__main__":
    main()

```

**Comment by leseb** [comment]
No activity. Closing this, feel free to re-open.

**Comment by mcv21** [comment]
@leseb do you think this should be fixed in stable-2.2?

**Comment by leseb** [comment]
@mcv21 stable-2.3 is around the corner, but if we have a fix in master we can backport it to 2.2.

**Comment by mcv21** [comment]
@leseb Sorry, I was obviously unclear. I was asking if you thought that the underlying bug had been fixed in stable-2.2?
